File: solutions_2755486_0/Python/hackedd/wall.py
import sys
from collections import namedtuple

def read_int(fp=sys.stdin):
    return int(fp.readline().strip())

def read_ints(fp=sys.stdin):
    return map(int, fp.readline().strip().split())

MIN_W = -100
MAX_W = 100
OFFSET_W = 200

class Tribe:
    def __init__(self, day, number, west, east, strength, delta_d, delta_p, delta_s):
        self.day = day            # the day of the tribe's first attack (where 1st January, 250BC, is considered day 0)
        self.number = number      # the number of attacks from this tribe
        self.west = west          # the westmost and eastmost points respectively of the Wall attacked on the first attack
        self.east = east
        self.strength = strength  # the strength of the first attack
        self.delta_d = delta_d    # the number of days between subsequent attacks by this tribe
        self.delta_p = delta_p    # the distance this tribe travels to the east between subsequent attacks (if this is negative, the tribe travels to the west)
        self.delta_s = delta_s    # the change in strength between subsequent attacks

def solve(tribes):
    wall = [0] * (MAX_W * 4)
    success = 0
    day = 0

    while day < 676060:
        next_day = None
        improve = []

        # print >>sys.stderr, "Day:", day

        for t, tribe in enumerate(tribes, 1):
            if tribe.day == day and tribe.number > 0:
                # print >>sys.stderr, "Tribe", t, "[", tribe.west, ",", tribe.east, "] strength =", tribe.strength
                win = False
                for i in range(tribe.west, tribe.east+1):
                    if wall[OFFSET_W + i] < tribe.strength:
                        improve.append((i, tribe.strength))
                        win = True
                        # print >>sys.stderr, "Succeeds at", i, "height =", wall[OFFSET_W + i]

                if win:
                    success += 1

                tribe.number -= 1
                tribe.day += tribe.delta_d
                tribe.west += tribe.delta_p
                tribe.east += tribe.delta_p
                tribe.strength += tribe.delta_s

            if tribe.number > 0 and (next_day is None or tribe.day < next_day):
                next_day = tribe.day

        if next_day is None:
            break
        day = next_day

        for i, s in improve:
            # Multiple attacks on same day!
            # Make sure not to raise and then lower again!
            if wall[OFFSET_W + i] < s:
                wall[OFFSET_W + i] = s

    return success

if __name__ == "__main__":
    T = read_int()
    for t in range(1, T+1):
        N = read_int()
        tribes = []
        for i in range(N):
            tribes.append(Tribe(*read_ints()))

        print "Case #%d: %s" % (t, solve(tribes))

File: solutions_2755486_0/C++/marcalla/wall.cpp
#include <cstdio>
#include <vector>
#include <algorithm>

using namespace std;

struct attack {
  int day;
  int strength;
  int west;
  int east;

  attack (int d, int s, int w, int e) : day(d), strength(s), west(w), east(e)  {};
};

struct stone {
  int height;
  int candidate;
};

bool compareAttacks (attack a, attack b) {
  if (a.day != b.day)
    return (a.day < b.day);
  else
    return (a.strength < b.strength);
}

int main () {

  stone heights [1001]; // center: 500
  int center = 500;
  int T;
  scanf ("%d", &T);
  for (int t = 1; t <= T; ++t) {
    vector <attack> attacks;
    for (int i = 0; i < 1001; ++i) {
      heights[i].height = 0;
      heights[i].candidate = 0;
    }
    int ntribes;
    scanf ("%d", &ntribes);
    //printf ("ntribes = %d\n", ntribes);
    for (int j = 0; j < ntribes; ++j) {
      int d, n, w, e, s, deltaD, deltaP, deltaS;
      scanf ("%d%d%d%d%d%d%d%d", &d, &n, &w, &e, &s, &deltaD, &deltaP, &deltaS);
      for (int i = 0; i < n; ++i) {
	attacks.push_back(attack(d, s, w, e));
	d += deltaD;
	w += deltaP;
	e += deltaP;
	s += deltaS;
      }
    }
    sort (attacks.begin(), attacks.end(), compareAttacks);
    //for (int i = 0; i < (int)attacks.size(); ++i)
    //printf ("day: %d strength: %d west: %d east: %d\n", attacks[i].day, attacks[i].strength, attacks[i].west, attacks[i].east);

    int successful = 0;
    for (int i = 0; i < (int)attacks.size(); ++i) {
      if (i != 0 && attacks[i].day != attacks[i-1].day) {
	for (int j = 0; j < 1001; ++j)
	  heights[j].height = heights[j].candidate;
      }
      bool tSuc = false;
      for (int j = attacks[i].west*2 + center; j <= attacks[i].east*2+center; ++j) {
	if (attacks[i].strength > heights[j].height) {
	  //printf ("attack #%d succeeds!\n", i);
	  tSuc = true;
	  heights[j].candidate = max(heights[j].candidate, attacks[i].strength);
	}
      }
      if (tSuc)
	++successful;
    }
    printf ("Case #%d: %d\n", t, successful);
  }

  return 0;
}
